src/DataStructures/Tensor/Expressions/Product.hpp

```diff
@@ -151,12 +151,12 @@ struct OuterProduct<T1, T2, IndexList1<Indices1...>, IndexList2<Indices2...>,
 /// \param t2 the second operand expression of the product
 /// \return the tensor expression representing the product of two tensor
 /// expressions
-template <typename T1, typename T2, typename ArgsList1, typename ArgsList2>
+template <typename T1, typename T2, typename X1, typename X2, typename Symm1,
+          typename Symm2, typename IndexList1, typename IndexList2,
+          typename ArgsList1, typename ArgsList2>
 SPECTRE_ALWAYS_INLINE auto operator*(
-    const TensorExpression<T1, typename T1::type, typename T1::symmetry,
```

Comment:
I'm not 100% *why* this is the case, but it turns out that without this change, deduced types for the templates do not work as expected and produce a compile error. Because I was seeing this error with products and not addition and I noticed that `operator+()` and `operator-()` in `AddSubtract.hpp` were templated differently, I just updated `operator*()` to be templated in the same way, and it fixed the error. I'm not 100% sure why the type deduction doesn't work properly without this change.

I'll compile the code locally without this change to `operator*()` for a test expression and will post the output here

Reply:
**Short response:** this is not necessary and this change masked the real problem (the base `TensorExpression` symmetry for CRTP with `TensorAsExpression` needed to match the derived type symmetry), which is fixed in the new fixup de0d574

**Explanation:**
I locally rolled back one commit to this 2nd commit, removed this specific change to `operator+()`, and compiled the `TensorExpression` unit tests. This is the test case in `Test_Product.cpp` that yields the error without this change to `operator+()`:
```
  // \f$L_{j, a, i}{}^{b} = G_{i, a} * H_{j}{}^{b}\f$
  // rank 2 x rank 2 outer product containing a spacetime index with a generic
  // spatial index
  const Tensor<DataType, Symmetry<4, 3, 2, 1>,
               index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpatialIndex<3, UpLo::Lo, Frame::Inertial>,
                          SpacetimeIndex<3, UpLo::Up, Frame::Inertial>>>
      GH = TensorExpressions::evaluate<ti_j, ti_a, ti_i, ti_B>(G(ti_i, ti_a) *
                                                               H(ti_j, ti_B));
```

The error seems to be saying that, for this test case, the deduced type for the first argument (first argument to `operator*()`, which is `G(ti_i, ti_a)`) is `TensorExpression<TensorAsExpression<Tensor<DataType, Symmetry<1, 1>, ...>, DataType, Symmetry<2, 1>, ...>>`, does not match the adjusted type, which is: `TensorAsExpression<Tensor<DataType, Symmetry<1, 1>, ...>`. Essentially, something with the CRTP is wrong since the the base symmetry is `<2, 1>` but the derived symmetry is `<1, 1>`. I believe that this was happening because in the 2nd commit when I added `TensorAsExpressionSymm` to compute the new "result" symmetry, I forgot to also use this new computed symmetry for the base type for CRTP. See the fixup commit (de0d574), where I fix the symmetry in the base type for CRTP and undo the unnecessary change to `operator()*`. I can squash once you've been able to see it and make sure it makes sense.